lxd/device/unix_hotplug.go

```diff
@@ -149,31 +149,42 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{}
 	runConf.PostHooks = []func() error{d.Register}
 
-	device := d.loadUnixDevice()
-	if d.isRequired() && device == nil {
-		return nil, fmt.Errorf("Required Unix Hotplug device not found")
+	devices := d.loadUnixDevices()
+	if d.isRequired() && len(devices) <= 0 {
+		return nil, fmt.Errorf("Required unix hotplug device %q not found", d.name)
 	}
 
-	if device == nil {
-		return &runConf, nil
-	}
+	// Setup reverter.
+	revert := revert.New()
+	defer revert.Fail()
 
-	devnum := device.Devnum()
-	major := uint32(devnum.Major())
-	minor := uint32(devnum.Minor())
+	for _, device := range devices {
+		devnum := device.Devnum()
+		major := uint32(devnum.Major())
+		minor := uint32(devnum.Minor())
+
+		// Setup device.
+		var err error
+		if device.Subsystem() == "block" {
+			err = unixDeviceSetupBlockNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+		} else {
+			err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+		}
 
-	// setup device
-	var err error
-	if device.Subsystem() == "block" {
-		err = unixDeviceSetupBlockNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	} else {
-		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	}
+		// Remove unix device on failure to setup device.
+		revert.Add(func() {
+			err = unixDeviceRemove(d.inst.DevicesPath(), "unix", d.name, "", &runConf)
+			if err != nil {
+				d.logger.Error("Failed to remove unix device", logger.Ctx{"device": d.name, "err": err})
```

Comment:
Perhaps "Failed to revert..." or "Failed to clean up..." would be clearer here, as otherwise users may think the LXD instance device itself was being removed even though it technically never got started. Also "removed" indicates the device was trying to be "added", but that is not whats happening from a user perspective, the device is actually being "started".

Reply:
Agreed.